### PR TITLE
shadow: update to 4.8.1

### PR DESCRIPTION
**Pull request by neheb**
Fixed license information.

Removed patch requiring autoreconf and replaced with a configure variable.

Removed faulty patch that broke systems without a disabled crypt size hack.
Replaced with using a SED command as well as bcrypt, which works in musl.

Removed su patch and converted it to a SED command in the Makefile.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

Maintainer: nobody
Compile tested: malta

Fixes https://github.com/openwrt/packages/issues/11135

### CODE REVIEW

File: utils/shadow/patches/005-set-encrypt-method-sha512.patch [hunk 1]
```diff
@@ -1,11 +0,0 @@
---- a/etc/login.defs
-+++ b/etc/login.defs
-@@ -317,7 +317,7 @@ CHFN_RESTRICT		rwh
- # Note: If you use PAM, it is recommended to use a value consistent with
- # the PAM modules configuration.
- #
--#ENCRYPT_METHOD DES
-+ENCRYPT_METHOD SHA512
```
Comment (BKPepe) [inline utils/shadow/patches/005-set-encrypt-method-sha512.patch]: I don't like this. ping @Cynerd

File: utils/shadow/patches/005-set-encrypt-method-sha512.patch [hunk 2]
```diff
@@ -1,11 +0,0 @@
---- a/etc/login.defs
```
Comment (BKPepe) [inline utils/shadow/patches/005-set-encrypt-method-sha512.patch]: I don't like this. ping @Cynerd

Comment (neheb) [reply utils/shadow/patches/005-set-encrypt-method-sha512.patch]: Note that this patch is replaced by one of the SED lines.

**Comment by Cynerd** [comment]
In general I have nothing against this with exception that shadow should prefer stronger cipher if available. I am not suggesting to patch it for it, I just suggest to replace specified algorithm in default with SHA512 is available and use BCRYPT otherwise. My idea is to select default cipher depending on configuration. With this the sed is absolutely ideal over my patch so I support this change.

Do you think that adding detection on `CONFIG_MUSL_DISABLE_CRYPT_SIZE_HACK` symbol and setting appropriate cipher that way would be possible?

**Comment by neheb** [comment]
On second thought, yes.

Note that bcrypt is a stronger algorithm than sha512. That being said, it's quite slow on malta. Reducing the rounds would help I think. OTOH, bcrypt uses more memory than md5 or sha512crypt, which could be a problem for low memory devices.

Turns out that I tested this with the crypt size hack disabled. I will retest with it enabled.
Worst case scenario, I change the SED command to be conditional on CONFIG_MUSL_DISABLE_CRYPT_SIZE_HACK

Here are some actual benchmarks:

```
Hashtype: descrypt, DES(Unix), Traditional DES

Speed.Dev.#1.:   906.7 MH/s (95.05ms)
Speed.Dev.#2.:   897.1 MH/s (96.18ms)
Speed.Dev.#3.:   914.2 MH/s (95.81ms)
Speed.Dev.#4.:   912.1 MH/s (96.09ms)
Speed.Dev.#5.:   914.8 MH/s (95.75ms)
Speed.Dev.#6.:   908.5 MH/s (96.39ms)
Speed.Dev.#7.:   916.7 MH/s (95.59ms)
Speed.Dev.#8.:   920.3 MH/s (95.19ms)
Speed.Dev.#*.:  7290.6 MH/s

Hashtype: md5crypt, MD5(Unix), FreeBSD MD5, Cisco-IOS MD5

Speed.Dev.#1.:  9918.1 kH/s (97.10ms)
Speed.Dev.#2.:  9830.1 kH/s (97.70ms)
Speed.Dev.#3.: 10017.4 kH/s (97.36ms)
Speed.Dev.#4.:  9986.5 kH/s (96.92ms)
Speed.Dev.#5.:  9976.8 kH/s (97.74ms)
Speed.Dev.#6.:  9889.2 kH/s (97.29ms)
Speed.Dev.#7.: 10010.5 kH/s (97.40ms)
Speed.Dev.#8.: 10083.0 kH/s (96.94ms)
Speed.Dev.#*.: 79711.6 kH/s

Hashtype: bcrypt, Blowfish(OpenBSD)

Speed.Dev.#1.:    13094 H/s (47.51ms)
Speed.Dev.#2.:    13076 H/s (47.63ms)
Speed.Dev.#3.:    13292 H/s (46.88ms)
Speed.Dev.#4.:    13250 H/s (47.17ms)
Speed.Dev.#5.:    13296 H/s (46.86ms)
Speed.Dev.#6.:    13160 H/s (47.30ms)
Speed.Dev.#7.:    13260 H/s (46.98ms)
Speed.Dev.#8.:    13280 H/s (46.80ms)
Speed.Dev.#*.:   105.7 kH/s

Hashtype: sha256crypt, SHA256(Unix)

Speed.Dev.#1.:   388.8 kH/s (93.91ms)
Speed.Dev.#2.:   388.6 kH/s (93.97ms)
Speed.Dev.#3.:   386.2 kH/s (95.45ms)
Speed.Dev.#4.:   391.1 kH/s (93.37ms)
Speed.Dev.#5.:   386.0 kH/s (95.51ms)
Speed.Dev.#6.:   387.6 kH/s (94.22ms)
Speed.Dev.#7.:   390.5 kH/s (95.08ms)
Speed.Dev.#8.:   391.1 kH/s (95.10ms)
Speed.Dev.#*.:  3110.0 kH/s

Hashtype: sha512crypt, SHA512(Unix)

Speed.Dev.#1.:   147.5 kH/s (103.12ms)
Speed.Dev.#2.:   138.1 kH/s (102.84ms)
Speed.Dev.#3.:   148.0 kH/s (102.83ms)
Speed.Dev.#4.:   148.3 kH/s (102.57ms)
Speed.Dev.#5.:   148.9 kH/s (102.19ms)
Speed.Dev.#6.:   138.5 kH/s (102.55ms)
Speed.Dev.#7.:   149.0 kH/s (102.09ms)
Speed.Dev.#8.:   150.1 kH/s (101.38ms)
Speed.Dev.#*.:  1168.6 kH/s
```

**Comment by neheb** [comment]
Just tested glibc. bcrypt does not work at all. I'll change it to sha512 there.

**Comment by BKPepe** [comment]
What about [new shadow applets](https://gitlab.labs.nic.cz/turris/turris-build/blob/hbd/patches/packages/to-upstream/0022-shadow-update-to-version-4.8.patch)?

**Comment by neheb** [comment]
Why not.

**Comment by alisaeed** [comment]
Hi @neheb 
Please merge this change into 19.07 branch.
Thanks

**Comment by BKPepe** [comment]
Done. Cherry-picked.
https://github.com/openwrt/packages/commit/90e596693d896c229ccc0eaf51f8640c42e406ac

**Pull request merged**